File: backend/app/schemas/endorsement.py
"""批准（Endorsement）相关 Schema —— 难度4级时的最终审核"""
from datetime import datetime
from pydantic import BaseModel, Field


class EndorsementListItem(BaseModel):
    """批准列表项"""
    id: int
    instance_id: int
    instance_name: str = ""
    node_id: int
    node_name: str = ""
    task_id: int | None = None
    endorser_id: int
    status: str
    is_end_node: bool = False
    round: int = 1
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class EndorsementDetail(BaseModel):
    """批准详情 —— 用于批准操作页"""
    id: int
    instance_id: int
    instance_name: str = ""
    instance_status: str = ""
    initiator_id: int
    initiator_name: str = ""
    priority: str = "normal"
    difficulty: str = "1"
    node_id: int
    node_name: str
    node_status: str
    task_id: int | None = None
    endorser_id: int
    endorser_name: str = ""
    status: str
    opinion: str | None = None
    round: int = 1
    require_endorser_signature: bool = True
    signature_x: float = 400
    signature_y: float = 100
    signature_page: int = -1
    current_signature_url: str | None = None  # 当前批准人的签名图片URL
    current_node_index: int = 0
    total_nodes: int = 0
    nodes: list[dict] = []      # 节点简要列表（进度链）
    files: list[dict] = []      # 当前轮次文件
    checks: list[dict] = []     # 校验记录
    approvals: list[dict] = []  # 审批记录
    decided_at: datetime | None = None
    created_at: datetime | None = None


class EndorseRequest(BaseModel):
    """批准通过请求"""
    opinion: str | None = Field(None, max_length=500, description="批准意见")
    signatures: list[dict] | None = Field(None, description="签名列表 [{file_id, signature_x, signature_y, signature_page}]")
    signature_x: float | None = Field(None, description="签名X坐标（旧版兼容）")
    signature_y: float | None = Field(None, description="签名Y坐标（旧版兼容）")
    signature_page: int | None = Field(None, description="签名页码（旧版兼容）")


class EndorseRejectRequest(BaseModel):
    """批准驳回请求"""
    opinion: str = Field(..., min_length=1, max_length=500, description="驳回意见（必填）")

File: frontend/src/api/endorsement.ts
/** 批准 API —— 难度4级的最终审核环节 */
import request from './request'
import type { SignatureSlot } from './signature'

/** 批准列表项 */
export interface EndorsementListItem {
  id: number
  instance_id: number
  instance_name: string
  node_id: number
  node_name: string
  task_id: number | null
  endorser_id: number
  status: string
  is_end_node: boolean
  round: number
  deadline: string | null       // 截止时间（ISO 格式）
  is_overdue: boolean            // 是否逾期
  days_remaining: number | null  // 剩余天数（负数=已逾期）
  created_at: string | null
}

/** 批准详情 */
export interface EndorsementDetail {
  id: number
  instance_id: number
  instance_name: string
  instance_status: string
  initiator_id: number
  initiator_name: string
  priority: string
  difficulty: string
  node_id: number
  node_name: string
  node_status: string
  node_description: string | null     // 节点说明
  time_limit_days: number | null      // 完成时限（天）
  deadline: string | null             // 截止时间
  task_id: number | null
  endorser_id: number
  endorser_name: string
  status: string
  opinion: string | null
  round: number
  require_endorser_signature: boolean
  signature_x: number
  signature_y: number
  signature_page: number
  current_signature_url: string | null
  role_signature: { x: number; y: number } | null  // 角色维度签名默认坐标
  current_node_index: number
  total_nodes: number
  nodes: { id: number; name: string; is_start: boolean; is_end: boolean; status: string }[]
  files: { id: number; original_name: string; mime_type: string | null; file_size: number | null; round: number; node_id: number | null; node_name: string }[]
  node_files: { id: number; original_name: string; mime_type: string | null; file_size: number | null; round: number }[]  // 仅本节点文件（签批用，后端过滤）
  checks: { id: number; checker_id: number; checker_name: string; status: string; opinion: string | null; decided_at: string | null }[]
  approvals: { id: number; approver_id: number; approver_name: string; status: string; opinion: string | null; signature_applied: boolean; decided_at: string | null }[]
  decided_at: string | null
  created_at: string | null
}

/** 获取我的批准列表 */
export async function getEndorsements(params: { type?: string; status?: string; keyword?: string; page?: number; page_size?: number } = {}) {
  const res = await request.get('/endorsements', { params })
  return res.data as { items: EndorsementListItem[]; total: number; page: number; page_size: number }
}

/** 获取批准详情 */
export async function getEndorsementDetail(id: number): Promise<EndorsementDetail> {
  const res = await request.get(`/endorsements/${id}`)
  return res.data
}

/** 批准通过 —— 支持多签名 */
export async function endorseApprove(
  id: number,
  opinion?: string | null,
  signatures?: SignatureSlot[] | null,
) {
  const res = await request.post(`/endorsements/${id}/approve`, {
    opinion,
    signatures,
  })
  return res.data
}

/** 批准驳回 */
export async function endorseReject(id: number, opinion: string) {
  const res = await request.post(`/endorsements/${id}/reject`, { opinion })
  return res.data
}
